test/opt/fold_test.cpp

```diff
@@ -5214,7 +5254,97 @@ INSTANTIATE_TEST_SUITE_P(IntegerRedundantFoldingTest, GeneralInstructionFoldingT
+            "%3 = OpLoad %uint %n\n" +
+            "%2 = OpShiftRightArithmetic %uint %uint_0 %3\n" +
+            "OpReturn\n" +
+            "OpFunctionEnd",
+        2, 93),
+    // Test case 20: Fold 0 << n
+    InstructionFoldingCase<uint32_t>(
+        Header() + "%main = OpFunction %void None %void_func\n" +
+            "%main_lab = OpLabel\n" +
+            "%n = OpVariable %_ptr_uint Function\n" +
+            "%3 = OpLoad %uint %n\n" +
+            "%2 = OpShiftLeftLogical %uint %uint_0 %3\n" +
+            "OpReturn\n" +
+            "OpFunctionEnd",
+        2, 93),
+    // Test case 21: Fold 0 / n
+    InstructionFoldingCase<uint32_t>(
+        Header() + "%main = OpFunction %void None %void_func\n" +
+            "%main_lab = OpLabel\n" +
+            "%n = OpVariable %_ptr_uint Function\n" +
+            "%3 = OpLoad %uint %n\n" +
+            "%2 = OpSDiv %uint %uint_0 %3\n" +
+            "OpReturn\n" +
+            "OpFunctionEnd",
+        2, 93),
+    // Test case 22: Fold 0 / n
+    InstructionFoldingCase<uint32_t>(
+        Header() + "%main = OpFunction %void None %void_func\n" +
+            "%main_lab = OpLabel\n" +
+            "%n = OpVariable %_ptr_uint Function\n" +
+            "%3 = OpLoad %uint %n\n" +
+            "%2 = OpUDiv %uint %uint_0 %3\n" +
+            "OpReturn\n" +
+            "OpFunctionEnd",
+        2, 93),
+    // Test case 23: Fold 0 % n
+    InstructionFoldingCase<uint32_t>(
+        Header() + "%main = OpFunction %void None %void_func\n" +
+            "%main_lab = OpLabel\n" +
+            "%n = OpVariable %_ptr_uint Function\n" +
+            "%3 = OpLoad %uint %n\n" +
+            "%2 = OpSMod %int %int_0 %3\n" +
+            "OpReturn\n" +
+            "OpFunctionEnd",
+        2, 100),
+    // Test case 24: Fold 0 % n
+    InstructionFoldingCase<uint32_t>(
+        Header() + "%main = OpFunction %void None %void_func\n" +
+            "%main_lab = OpLabel\n" +
+            "%n = OpVariable %_ptr_uint Function\n" +
+            "%3 = OpLoad %uint %n\n" +
+            "%2 = OpUMod %uint %uint_0 %3\n" +
+            "OpReturn\n" +
+            "OpFunctionEnd",
+        2, 93),
+    // Test case 25: Fold n % 1
+    InstructionFoldingCase<uint32_t>(
+        Header() + "%main = OpFunction %void None %void_func\n" +
+            "%main_lab = OpLabel\n" +
+            "%n = OpVariable %_ptr_uint Function\n" +
+            "%3 = OpLoad %int %n\n" +
+            "%2 = OpSMod %int %3 %int_1\n" +
+            "OpReturn\n" +
+            "OpFunctionEnd",
+        2, 226),
+    // Test case 26: Fold n % 1
+    InstructionFoldingCase<uint32_t>(
+        Header() + "%main = OpFunction %void None %void_func\n" +
+            "%main_lab = OpLabel\n" +
+            "%n = OpVariable %_ptr_uint Function\n" +
+            "%3 = OpLoad %uint %n\n" +
+            "%2 = OpUMod %uint %3 %uint_1\n" +
+            "OpReturn\n" +
+            "OpFunctionEnd",
+        2, 226),
```

Comment:
Is there a way to dump the IR state to get what 226 is? Because it's not one of the predefined constants so I'm not 100% sure it's correct (expecting [u]int 0 in both), this will come from the synthesised constant 0 you suggested I think.

Reply:
I did not look at it too closely when reviewing. Those tests are not stable. The 226 is supposed to be the id of the resulting instruction. In this case, the result id is %uint_1, whose numeric id in the binary could easily change.

There are examples of tests that do stuff like that. Look for macros at the top of the file.

Reply:
https://github.com/KhronosGroup/SPIRV-Tools/blob/main/test%2Fopt%2Ffold_test.cpp#L202

Reply:
You can also move to a different type of test where you expect the result to be a constant. You give the value instead of the id of the instruction.

Reply:
Here are some that might be useful 

https://github.com/KhronosGroup/SPIRV-Tools/blob/main/test%2Fopt%2Ffold_test.cpp#L466
https://github.com/KhronosGroup/SPIRV-Tools/blob/main/test%2Fopt%2Ffold_test.cpp#L1134
https://github.com/KhronosGroup/SPIRV-Tools/blob/main/test%2Fopt%2Ffold_test.cpp#L1295

Reply:
Okay, apparently "226" was OpConstantNull which is why it wasn't merging. I've named (well numbered) all the constants now.